Added ByIgnoringPropertiesOfTypeAssignableTo extension method and refactored core to make it possible

chase-miller opened the pull request:
This adds a `ByIgnoringPropertiesOfTypeAssignableTo<T>()` extension method. My use case is that I want to ignore particular properties on all objects that are derived from a particular interface.

In order to do this I refactored the core to be more extensible such that others can write their own extension methods to ignore properties under other conditions.

sungam3r commented on src/Destructurama.ByIgnoring/Destructurama.ByIgnoring.csproj:
```diff
@@ -1,7 +1,7 @@
 
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0;netstandard1.0</TargetFrameworks>
+    <TargetFrameworks>netstandard2.0;netstandard1.1</TargetFrameworks>
```
@nblumhardt I would drop netstandard1.1

chase-miller replied:
This seems reasonable to me. If we do we might want to increment a major version since it's a breaking change. Although the constructor change to `DestructureByIgnoringPolicy` might warrant a major version bump on its own (I think it's public, right?).

sungam3r replied:
Another option - wait for Serilog v3 release updating dependencies and bumping major version here with all breaking changes you want.

chase-miller replied:
I mean I'd prefer to not wait for Serilog v3 ;-)

So forget I said anything haha

sungam3r replied:
@chase-miller Hi. I was away from this repo and all OSS for a long time but now I have time to proceed here. Please pull from dev and resolve conflicts. Then I'm fine to review/merge.

sungam3r commented on test/Destructurama.ByIgnoring.Tests/Destructurama.ByIgnoring.Tests.csproj:
```diff
@@ -1,7 +1,7 @@
 
   <PropertyGroup>
-    <TargetFrameworks>netcoreapp3.0;netcoreapp2.2;netcoreapp1.1</TargetFrameworks>
+    <TargetFrameworks>netcoreapp2.2;netcoreapp3.0;net6.0</TargetFrameworks>
```
```suggestion
    <TargetFrameworks>netcoreapp2.2;netcoreapp3.0;net7.0</TargetFrameworks>
```

sungam3r commented on src/Destructurama.ByIgnoring/LoggerConfigurationIgnoreExtensions.cs:
```diff
@@ -32,9 +34,46 @@ public static class LoggerConfigurationIgnoreExtensions
         /// <param name="ignoredProperty">The function expressions that expose the properties to ignore.</param>
         /// <returns>An object allowing configuration to continue.</returns>
-        public static LoggerConfiguration ByIgnoringProperties<TDestruture>(this LoggerDestructuringConfiguration configuration, params Expression<Func<TDestruture, object>>[] ignoredProperty)
+        public static LoggerConfiguration ByIgnoringProperties<TDestruture>(this LoggerDestructuringConfiguration configuration, params Expression<Func<TDestruture, object>>[] ignoredProperty) =>
+            configuration.ByIgnoringPropertiesWhere(obj => obj.GetType() == typeof(TDestruture), ignoredProperty);
```
```suggestion
        public static LoggerConfiguration ByIgnoringProperties<TDestructure>(this LoggerDestructuringConfiguration configuration, params Expression<Func<TDestructure, object>>[] ignoredProperty) =>
            configuration.ByIgnoringPropertiesWhere(obj => obj.GetType() == typeof(TDestructure), ignoredProperty);
```

sungam3r commented on src/Destructurama.ByIgnoring/LoggerConfigurationIgnoreExtensions.cs:
```diff
@@ -32,9 +34,46 @@ public static class LoggerConfigurationIgnoreExtensions
+        /// <param name="configuration">The logger configuration to apply configuration to.</param>
+        /// <param name="ignoredProperty">The function expressions that expose the properties to ignore.</param>
+        /// <returns>An object allowing configuration to continue.</returns>
+        public static LoggerConfiguration ByIgnoringPropertiesOfTypeAssignableTo<TDestruture>(this LoggerDestructuringConfiguration configuration, params Expression<Func<TDestruture, object>>[] ignoredProperty) =>
+            configuration.ByIgnoringPropertiesWhere(obj => obj is TDestruture, ignoredProperty);
```
```suggestion
        public static LoggerConfiguration ByIgnoringPropertiesOfTypeAssignableTo<TDestructure>(this LoggerDestructuringConfiguration configuration, params Expression<Func<TDestructure, object>>[] ignoredProperty) =>
            configuration.ByIgnoringPropertiesWhere(obj => obj is TDestructure, ignoredProperty);
```

sungam3r commented on src/Destructurama.ByIgnoring/LoggerConfigurationIgnoreExtensions.cs:
```diff
@@ -32,9 +34,46 @@ public static class LoggerConfigurationIgnoreExtensions
+
+        /// <summary>
+        /// Destructure.ByIgnoringProperties takes one or more expressions that access a property, e.g. obj => obj.Property, and uses the property names to determine which
+        /// properties are ignored when an object of type assignable to TDestruture is destructured by serilog.
```
```suggestion
        /// properties are ignored when an object of type assignable to TDestructure is destructured by serilog.
```

sungam3r commented on src/Destructurama.ByIgnoring/LoggerConfigurationIgnoreExtensions.cs:
```diff
@@ -32,9 +34,46 @@ public static class LoggerConfigurationIgnoreExtensions
+            configuration.ByIgnoringPropertiesWhere(obj => obj.GetType() == typeof(TDestruture), ignoredProperty);
+
+        /// <summary>
+        /// Destructure.ByIgnoringProperties takes one or more expressions that access a property, e.g. obj => obj.Property, and uses the property names to determine which
```
```suggestion
        /// Destructure.ByIgnoringPropertiesOfTypeAssignableTo takes one or more expressions that access a property, e.g. obj => obj.Property, and uses the property names to determine which
```

sungam3r commented on src/Destructurama.ByIgnoring/LoggerConfigurationIgnoreExtensions.cs:
```diff
@@ -32,9 +34,46 @@ public static class LoggerConfigurationIgnoreExtensions
+            configuration.ByIgnoringPropertiesWhere(obj => obj is TDestruture, ignoredProperty);
+
+        /// <summary>
+        /// Destructure.ByIgnoringProperties takes one or more expressions that access a property, e.g. obj => obj.Property, and uses the property names to determine which
```
```suggestion
        /// Destructure.ByIgnoringPropertiesWhere takes one or more expressions that access a property, e.g. obj => obj.Property, and uses the property names to determine which
```

sungam3r commented on src/Destructurama.ByIgnoring/LoggerConfigurationIgnoreExtensions.cs:
```diff
@@ -32,9 +34,46 @@ public static class LoggerConfigurationIgnoreExtensions
+
+        /// <summary>
+        /// Destructure.ByIgnoringProperties takes one or more expressions that access a property, e.g. obj => obj.Property, and uses the property names to determine which
+        /// properties are ignored when an object for which destructureFunc returns true is destructured by serilog.
```
```suggestion
        /// properties are ignored when an object for which handleDestructuringPredicate returns true is destructured by serilog.
```

sungam3r commented on src/Destructurama.ByIgnoring/LoggerConfigurationIgnoreExtensions.cs:
```diff
@@ -32,9 +34,46 @@ public static class LoggerConfigurationIgnoreExtensions
         }
+
+        /// <summary>
+        /// Destructure.ByIgnoringProperties takes one or more ignoredProperty predicates that when true indicates a given property is to be ignored when destructured by serilog.
```
```suggestion
        /// Destructure.ByIgnoringPropertiesWhere takes one or more ignoredProperty predicates that when true indicates a given property is to be ignored when destructured by serilog.
```

chase-miller commented:
@sungam3r thank you for all the feedback! I plan to go through it and accept all the recommended changes aside from the dropping of netstandard1.0/1 today (wait until we get feedback on it). I should probably write some more unit tests, so I hope to do that today as well.

sungam3r commented on src/Destructurama.ByIgnoring/LoggerConfigurationIgnoreExtensions.cs:
```diff
@@ -32,9 +34,46 @@ public static class LoggerConfigurationIgnoreExtensions
         /// <param name="ignoredProperty">The function expressions that expose the properties to ignore.</param>
         /// <returns>An object allowing configuration to continue.</returns>
-        public static LoggerConfiguration ByIgnoringProperties<TDestruture>(this LoggerDestructuringConfiguration configuration, params Expression<Func<TDestruture, object>>[] ignoredProperty)
+        public static LoggerConfiguration ByIgnoringProperties<TDestructure>(this LoggerDestructuringConfiguration configuration, params Expression<Func<TDestructure, object>>[] ignoredProperty) =>
```
```suggestion
        public static LoggerConfiguration ByIgnoringProperties<TDestructure>(this LoggerDestructuringConfiguration configuration, params Expression<Func<TDestructure, object>>[] ignoredProperties) =>
```

and in all other methods I think

sungam3r commented on src/Destructurama.ByIgnoring/LoggerConfigurationIgnoreExtensions.cs:
```diff
@@ -32,9 +34,46 @@ public static class LoggerConfigurationIgnoreExtensions
         /// <param name="configuration">The logger configuration to apply configuration to.</param>
         /// <param name="ignoredProperty">The function expressions that expose the properties to ignore.</param>
```
```suggestion
        /// <param name="ignoredProperties">The array of function expressions that expose the properties to ignore.</param>
```

chase-miller commented:
@sungam3r I applied all suggested changes. I also added a bunch of tests. It got to the point where there was a lot of repetition, so I reworked the tests to be data-driven. This got rid of some of the repetition but not all (notice the comment on `ByIgnoringPropertiesOfTypeAssignableToTestCases.cs:13`).

Let me know what you think; I'm happy to update based on your feedback. And thank you!

sungam3r commented on src/Destructurama.ByIgnoring/ByIgnoring/IgnoredPropertyExpressionExtensions.cs:
```diff
@@ -17,11 +17,17 @@
+    /// <summary>
+    /// Extension methods used to obtain property names.
+    /// </summary>
+    public static class IgnoredPropertyExpressionExtensions
```
Whe public?

chase-miller replied:
I made it public so that the test project could access the method (`ByIgnoreWhereTestCases.cs:125`). I'm not a huge fan of this to be honest (it was an effort to reduce what are essentially duplicate test case definitions), so if it's a deal-breaker I can come up with another solution and make this internal again.

sungam3r replied:
InternalsVisibleTo is a well known way for testing.

sungam3r replied:
fixed in #25

sungam3r commented on test/Destructurama.ByIgnoring.Tests/Destructurama.ByIgnoring.Tests.csproj:
```diff
@@ -1,13 +1,14 @@
 
   <PropertyGroup>
-    <TargetFrameworks>netcoreapp3.0;netcoreapp2.2;netcoreapp1.1</TargetFrameworks>
     <LangVersion>latest</LangVersion>
```
```suggestion
```

chase-miller replied:
Just updated

sungam3r commented on test/Destructurama.ByIgnoring.Tests/TestCases/ByIgnoringTestCases.cs:
```diff
@@ -0,0 +1,202 @@
+// (C) Copyright 2023 Nullable, Inc. All rights reserved.
+// This file is part of Nullable's product Aware and cannot be copied and/or
+// distributed without the express permission of Nullable, Inc.
```
?

chase-miller replied:
This was accidentally added. Just pushed a commit that replaces these with the proper copyright notice.

chase-miller commented on test/Destructurama.ByIgnoring.Tests/Destructurama.ByIgnoring.Tests.csproj:
```diff
@@ -1,13 +1,14 @@
-    <TargetFrameworks>netcoreapp3.0;netcoreapp2.2;netcoreapp1.1</TargetFrameworks>
     <LangVersion>latest</LangVersion>
     <RootNamespace>Destructurama.ByIgnoring.Tests</RootNamespace>
     <AssemblyName>Destructurama.ByIgnoring.Tests</AssemblyName>
```
Just updated

chase-miller commented on src/Destructurama.ByIgnoring/ByIgnoring/IgnoredPropertyExpressionExtensions.cs:
```diff
@@ -17,11 +17,17 @@
                                                       "Only function expressions that access a property are currently supported. e.g. obj => obj.Property";
 
+        /// <summary>
+        /// Obtains the name of a property of the provided TDestructureType using an expression.
```
Good call. But adding that results in a compilation error (CS1574), and I haven't been able to figure out how to resolve.

sungam3r replied:
fixed in #25

nblumhardt commented:
Hi! Sorry to have taken a while to chime in, @chase-miller, and thanks for the PR.

I think the goals of this library are pretty modest, and we haven't been able to dedicate a lot of maintenance time to it, so we've also kept the amount of code and surface area very small.

I think for this situation, either writing a custom `IDestructuringPolicy` for your app and scenario, or publishing a forked version of this project under a different name, would be the way to go.

sungam3r commented:
@nblumhardt I can help to support this repo on an ongoing basis.

nblumhardt commented:
Invited- thank you @sungam3r 🙇

sungam3r commented:
@chase-miller Please resolve conflicts/rebase onto `dev`.

sungam3r commented:
Superseded by #25.
